### ISSUE TITLE
Automated end to end test

### ISSUE THREAD

**Issue by ewjoachim** [ewjoachim/python-coverage-comment-action#29]
Ping @kieferro

Trying to summarize what automated end to end tests would look like:
- I think we'd need personnal access tokens from 2 distinct users. We can create 2 dedicated bot users just for the CI. The tests will be automatically skipped if the 2 tokens aren't passed via the env var
- We'll use pytest fixtures to handle the creation and deletion of resources.
- We'll use pytest fixtures to ease calling the `gh` CLI. The GH CLI can often output json, so we'll be able to easily read and write info to and from GitHub
- And then we'll need some other fixtures:
- `repository` fixture
  - `cd` to a temp directory
  - write a git repo with some code and the action configured (as well as a CI that computes coverage on some code)
  - create a github repo and push to it (wiki is active by default)
  - After the test, delete the github repo
- `fork` fixture
  - fork the repo (bonus points for forking with a different repo name :) )
  - clone the fork
  - After the test, delete the fork
- `pr` fixture (can be run from either user)
  - checkout a new branch
  - change some code
  - git commit -am foo && gh pr open

Also, `gh run watch --exit-status` lets us easily have a command we can call that will wait for the CI to run and raise if it failed.

With all that we can have:
- A test that pushes some code on the default branch, waits for the CI and checks that the Coverage Badge is working.
- A test that pushes some code, then creates a branch and pushes code from the same user, check that the comment is posted
- A test that pushes some code, then from a different user, pushes code in a new branch and check that the comment is also posted

If we have that and given we have 100% coverage, we should be quite all set :D

**Comment by ewjoachim** [comment]
I've made a PR that approximate the text above in very very early untested stage. Let's iterate.

**Comment by ewjoachim** [comment]
I've added you to the repo, so you should be able to push on the branch :)

**Comment by ewjoachim** [comment]
Ha.

I've added some code to the PR but I've arrived to a road block. For this action to work, we need the wiki, which needs to be initialized with a first page and it look like there isn't a way to initialize the wiki repo automatically.

Not spending longer on that PR unless we find a solution.

**Comment by ewjoachim** [comment]
Only 2 solutions I see so far are "selenium-like" interaction or asking a human to login and activate the wiki. I hate both solutions.

**Comment by kieferro** [comment]
> For this action to work, we need the wiki, which needs to be initialized with a first page and it look like there isn't a way to initialize the wiki repo automatically.

Yes, that is a problem. It should be possible to add a wiki via the Github API, but I haven't found a solution for creating the first page either.





> I don't like either solution.

Yes, I agree.

**Comment by kieferro** [comment]
I know that adds extra work and doesn't quite solve the original problem either. But I also wondered before if it might not be possible to store the files that are currently stored in the wiki in a gist (similar to how this action stores the values for dynamic badges in a gist:https://github.com/Schneegans/dynamic-badges-action).
This would have the advantage that also private repos can have a badge, because the gist is public.
Of course, the default should still be to use the wiki, so that none of the existing use cases are broken. But there could be the possibility to choose one of the two ways by an optional parameter.

Of course, that in itself is an idea that has nothing to do with this issue, but I mention it here because it would theoretically solve the problem with the wiki if end-to-end testing is then simply done with a gist. Of course, in that case, the end-to-end testing would not test the part where the data is shared via the wiki, but all other steps would be tested, which would at least avoid a very large part of the errors.

It's just an idea, let me know what you think.

**Comment by ewjoachim** [comment]
Have you seen this discussion ? :D https://github.com/Schneegans/dynamic-badges-action/issues/7

Since then I noticed that Gists can be created automatically (while the action you mentions requests the user to manually create the Gist). And also, at that point, I had not realized that Wikis required shenanigans to setup. Which means that gists would allow the action to work without shenanigans.

But, and it's an important point: I wanted the action to work without having to generate a PAT (which is another type of shenanigans). Also, if we require a PAT, then it means that the end to end tests would copy the PAT they use for creating the repo as a secret of the repo itself that would be used for creating gists. I guess it would work as an automated way.

Another issue would be that we need a way to link the gist and the repo. We could either use a specific string as a description and then search through all the user's gists (but it would likely be annoying), or store the gist ID somewhere (where ?). If the gist ID was saved as a variable of the workflow, it would likely mean that the creation of the gist could not be automated as part of the action, which would defeat the purpose of trying to avoid manual shenanigans to create the action (though I admit, it would be feasible to automate that in the end to end test by rewriting the workflow yaml file with the gist id)

If we could solve these issues, I'd be almost convinced that dropping wikis altogether and switching to Gists would be beneficial, and if we do a v3, then it won't break for existing v2 users (if I'm not mistaken, the only part that would break would be that we'd need to update the badge link)

To summarize:
- We'd like to avoid manual steps when setting up the action.
- If there are manual steps, we need those manual steps to be automatable (even if in an ugly way) for the end to end tests.

And then:
- Wiki has simple manual steps (activate the wiki, create a first page, opt. deactivate the wiki) when setting up the action, but no way of making those steps automated for the end to end tests
- Gist has more complex manual steps (create a secret gist, copy ID, paste ID in workflow, create PAT with gist scope, save as a secret) but that could be completely automated except for the PAT, and for the end to end tests, it could be 100% automated (by reusing a PAT we already have)

**Comment by ewjoachim** [comment]
Oh, btw, regarding the issue of private repos: I think it would be better solved by the action making a `requests.get(f"https://img.shields.io/badge/coverage-100%25-brightgreen")` and saving the resulting svg instead to our storage instead of the json (and have the "read" part" be able to extract coverage from the svg. Or use a different file altogether, which would have other advantages). The only reason I didn't do it this way first is that the Shields JSON endpoints allow easy customization of the badge, whereas this solution would make the badge not customizable.

**Comment by ewjoachim** [comment]
:o 

It actually works with Shields.io Dynamic badge ! No idea why the json endpoint doesn't work but the dynamic works.

![coverage 3](https://img.shields.io/badge/dynamic/json?color=brightgreen&label=coverage&query=%24.message&url=https%3A%2F%2Fgithub.com%2Fewjoachim%2Ftest-coverage-release%2Freleases%2Fdownload%2F1.0.0%2Fpython-coverage-comment-action-badge.json)

**Comment by kieferro** [comment]
> Have you seen this discussion ?

No, actually I didn't. It's funny that we both stumble across this topic/repo independently of each other ^^

> I wanted the action to work without having to generate a PAT

Yes, this is one of the big disadvantages of the Gist solution

> If the gist ID was saved as a variable of the workflow, it would likely mean that the creation of the gist could not be automated as part of the action

I thought it might be possible for the user to specify a name of the gist when writing the yaml file for the action. This name doesn't necessarily have to belong to a gist that already exists. The action would simply check if a gist with that name exists (this would probably require looping through all the gists, which would be a bit annoying). If no gist with this name exists yet, one will be created, otherwise the content of the existing one will be modified.
Is there any detail I'm not thinking of that would make this impossible?

> or it could mean that the action is able to edit its own workflow file or somehow commit a file to the repo, but I hate this idea

This would probably lead to very weird edge cases that don't work because the action first has to understand at which point the gist ID has to be specified. And if for example someone uses the action multiple times in a YAML file (I know that wouldn't make much sense, but I can't think of a better example and theoretically you would still expect it to work) weird things could happen because the gist ID is inserted in the wrong place.
Also, I agree that we should generally avoid having the action write anything to the repo content itself.

> We need ideas for new columns :)

Just as an idea: Automatically create an empty branch in the first run, where only the SVG for the badge and the JSON file are stored. The files are then changed each time the action is executed. This would also meet all the criteria. However, the action would have to create commits in this branch, which is also not ideal.

> The release would be visible in the list of releases

If it was marked as pre-release, that might not be so bad. This makes it relatively clear that it is not an official release and, for example, the link to the latest release (https://github.com/user/repo/releases/latest/...) only refers to real releases.
It would still be in the list, of course.

> I think it would be better solved by the action making a `requests.get(f"https://img.shields.io/badge/coverage-100%25-brightgreen")` and saving the resulting svg to our storage instead of the json

Yes, that makes sense. That would solve the problem with private repos for any solution.

> It actually works with Shields.io Dynamic badge ! No idea why the json endpoint doesn't work but the dynamic badge works.

Yes, that is very strange.

I actually really like the idea of releases. But I think we have some problems with the _Exposed in a predictable & stable URL_ aspect, because the files can't be accessed with raw.githubusercontent.com/...

**Comment by ewjoachim** [comment]
> That would solve the problem with private repos for any solution.

Strange thing: as shown in https://github.com/ewjoachim/test-coverage-release/blob/master/README.md, the svg-as-release-artifact produces a link that GitHub fails to use as an `img` target
<img width="1143" alt="Capture d’écran 2022-07-16 à 20 50 53" src="https://user-images.githubusercontent.com/1457576/179368330-66653d45-2e3a-4ae4-9674-651234117435.png">
<img width="301" alt="Capture d’écran 2022-07-16 à 20 51 04" src="https://user-images.githubusercontent.com/1457576/179368332-aa6b1a0d-943b-478b-aca3-96a88c3bb89d.png">

**Comment by ewjoachim** [comment]
> That would solve the problem with private repos for any solution.

Strange thing: as shown in https://github.com/ewjoachim/test-coverage-release/blob/master/README.md, the svg-as-release-artifact produces a link that GitHub fails to use as an `img` target
<img width="1143" alt="Capture d’écran 2022-07-16 à 20 50 53" src="https://user-images.githubusercontent.com/1457576/179368330-66653d45-2e3a-4ae4-9674-651234117435.png">
<img width="301" alt="Capture d’écran 2022-07-16 à 20 51 04" src="https://user-images.githubusercontent.com/1457576/179368332-aa6b1a0d-943b-478b-aca3-96a88c3bb89d.png">

No idea why but it may be because:
- The link is a redirect to `https://objects.githubusercontent.com/github-production-release-asset-` (doubt it would be the reason)
- The content-type of the response is `octet-stream` and not an image (doubt it would be the reason)
- For some reason, when inspecting the response in my browser for this call, I receive a base64 of the svg, meanwhile on curl, I get the SVG itself (could be a reason but I don't understand where that's from)
- github's "camo" mecanism (serves all images from a proxy) could be doing something funky

**Comment by ewjoachim** [comment]
> I actually really like the idea of releases. But I think we have some problems with the Exposed in a predictable & stable URL aspect, because the files can't be accessed with raw.githubusercontent.com/...

I'm not sure I understand what you mean. If the url is stable AND predictable AND on the github domain direction, how is that an issue ?

**Comment by ewjoachim** [comment]
Just confirmed: the "commit to a dedicated branch" solution works with direct svg upload, shields json endpoint and shields dynamic endpoint

**Comment by ewjoachim** [comment]
For the sake of it, I've also tried Draft releases:
- The URL is not predictable an changes all the time
- Plus the URL is only accessible for authenticated accounts that can access the draft release.

**Comment by ewjoachim** [comment]
On problem with "commit to a dedicated branch" though I belive it should be ok, is that it tends to display a banner like:
<img width="920" alt="Capture d’écran 2022-07-16 à 21 28 55" src="https://user-images.githubusercontent.com/1457576/179369405-1c5da729-6516-4d41-8cdd-c54a1c99578c.png">

Though I believe GitHub only suggests this for the branches pushed by you, and as long as the commit is made/pushed with the GHA token, I believe GitHub will not be displaying this to everyone... Not sure though.

**Comment by kieferro** [comment]
> I'm not sure I understand what you mean. If the url is stable AND predictable AND on the github domain direction, how is that an issue ? (apart from the fact shields.io blocks the github domain for some uses)

I worded that somewhat misleadingly. I was also referring to the problem you just raised: The link to the SVG file somehow does not really lead to the SVG file, so it cannot be used as an image link.

**Comment by kieferro** [comment]
> For the sake of it, I've also tried Draft releases

I'm not sure if this refers to my suggestion, but just in case it does: I was talking about pre-releases, which are simply labeled differently, not draft releases.

**Comment by kieferro** [comment]
> Though I believe GitHub only suggests this for the branches pushed by you, and as long as the commit is made/pushed with the GHA token, I believe GitHub will not be displaying this to everyone

Yes, I just tested it here (https://github.com/kieferro/test-dedicated-branch). I simply set up a script that adds a file and commits onto the `coverage` branch. I am only shown the branch that I have committed to myself. So this should not be a problem:

![Screenshot 2022-07-16 225539](https://user-images.githubusercontent.com/81773954/179371783-3f360fea-bd71-4eaa-84a9-7b2244b83155.png)


However, the Github home page shows that the bot pushed to my repo:

![Screenshot 2022-07-16 225554](https://user-images.githubusercontent.com/81773954/179371793-9395b84a-4f08-4e00-847f-99c101415f75.png)





> Just confirmed: the "commit to a dedicated branch" solution works with direct svg upload, shields json endpoint and shields dynamic endpoint

If the display on the homepage is not a problem, I think that would be a pretty good solution overall. What do you think?

**Comment by ewjoachim** [comment]
>  I was also referring to the problem you just raised: The link to the SVG file somehow does not really lead to the SVG file, so it cannot be used as an image link.

👍 

> I'm not sure if this refers to my suggestion, but just in case it does: I was talking about pre-releases, which are simply labeled differently, not draft releases.

Yeah, sorry it was my turn being unclear 😅 I know, it wasn't directly linked to your suggestion, but I figured out that if draft release files were accessible and the URLs were predictible, then it would solve most of the issues we had with releases.

**Comment by ewjoachim** [comment]
> If the display on the homepage is not a problem, I think that would be a pretty good solution overall. What do you think?

Hm, makes me wonder if it will probably also create notifications if you follow the repo.

If it doesn't I'm inclined to say let's go for it :)

**Comment by ewjoachim** [comment]
I think we've moved quite a lot from this issue's original topic. Let's make a new issue for discussing the specifics of changing the storage, as I believe doing that and adding end to end tests will be 2 distinct PRs

**Comment by kieferro** [comment]
> Yeah, sorry it was my turn being unclear 😅 I know, it wasn't directly linked to your suggestion, but I figured out that if draft release files were accessible and the URLs were predictible, then it would solve most of the issues we had with releases.

Okay, I figured that was your own idea, but I wasn't sure. All good :grinning:

**Comment by kieferro** [comment]
> Hm, makes me wonder if it will also create notifications if you follow the repo.

I think it's just like some person is creating commits on a branch. I usually don't get notifications for commits, but that may be due to my settings. I'm not quite sure what setup you are assuming. If you want, feel free to follow the repo and I'll trigger the workflow again to test that.

> If it doesn't I'm inclined to say let's go for it :)

If everything would work, that would be great :)

**Comment by kieferro** [comment]
> I think we've moved quite a lot from this issue's original topic.

Yes, you're right, we've gotten a bit off topic here :sweat_smile:

**Comment by ewjoachim** [comment]
> I think it's just like some person is creating commits on a branch.

I think you're right. You get notifications if there's a PR or if the commit message mentions you but a simple commit probably doesn't trigger a notification.
